packages/permissionless/accounts/kernel/signerToEcdsaKernelSmartAccount.ts

```diff
@@ -353,7 +535,11 @@ export async function signerToEcdsaKernelSmartAccount<
 
             if (smartAccountDeployed) return "0x"
 
-            return concatHex([factoryAddress, await generateInitCode()])
+            const _factoryAddress =
+                entryPointVersion === "v0.6"
+                    ? factoryAddress
+                    : metaFactoryAddress
```

Comment:
What is the difference between `metaFactoryAddress` and `factoryAddress` and why do we need two separate keys for it.

Reply:
In Kernel v3 we have new design for factory where there are two factories basically. 
The metaFactory is the one which is staked with the EntryPoint and the other actual factory is tied with implementation.

Considerations for this new design: 

- Before the single factory was tied with a single owner key.
- If owner key was compromised the attacker could allow a malicious implementation which can comprise existing accounts since changing implementation didn’t change counterfactual address.
- In new design only metaFactory is tied with a single owner key. Even if owner key is compromised the metaFactory only has the ability to approve a malicious factory which changes the counterfactual addresses anyway which doesn’t theoretically compromise existing accounts.
- The drawback of new design though is every new implementation will change the counterfactual address since there will be new factory as factory is tied with the implementation.
- Another plus side: One can do changes to implementation and how the factory works as well and still keep on using the same staking as before because metaFactory is the one which holds the stake with the EntryPoint for the bundlers.

Reply:
I see interesting.